lib/phoenix/transports/websocket.ex

```diff
@@ -13,36 +17,67 @@ defmodule Phoenix.Transports.WebSocket do
     ]
   end
 
-  def connect(%{method: "GET"} = conn, endpoint, handler, opts) do
+  def init(opts), do: opts
+
+  def call(%{method: "GET"} = conn, {endpoint, handler, opts}) do
     conn
-    |> Plug.Conn.fetch_query_params()
+    |> fetch_query_params()
     |> Transport.code_reload(endpoint, opts)
     |> Transport.transport_log(opts[:transport_log])
     |> Transport.force_ssl(handler, endpoint, opts)
     |> Transport.check_origin(handler, endpoint, opts)
     |> Transport.check_subprotocols(opts[:subprotocols])
     |> case do
       %{halted: true} = conn ->
-        {:error, conn}
+        conn
 
       %{params: params} = conn ->
         keys = Keyword.get(opts, :connect_info, [])
         connect_info = Transport.connect_info(conn, endpoint, keys)
-        config = %{endpoint: endpoint, transport: :websocket, options: opts, params: params, connect_info: connect_info}
+
+        config = %{
+          endpoint: endpoint,
+          transport: :websocket,
+          options: opts,
+          params: params,
+          connect_info: connect_info
+        }
+
+        cowboy_opts =
+          opts
+          |> Enum.flat_map(fn
+            {:timeout, timeout} -> [idle_timeout: timeout]
+            {:compress, _} = opt -> [opt]
+            {:max_frame_size, _} = opt -> [opt]
+            _other -> []
+          end)
+          |> Map.new()
+
+        process_flags =
+          opts
+          |> Keyword.take([:fullsweep_after])
+          |> Map.new()
 
         case handler.connect(config) do
-          {:ok, state} -> {:ok, conn, state}
-          :error -> {:error, Plug.Conn.send_resp(conn, 403, "")}
+          {:ok, state} ->
+            conn
+            |> upgrade_adapter(
+              :websocket,
+              {Phoenix.Endpoint.Cowboy2Handler, {handler, process_flags, state}, cowboy_opts}
+            )
```

Comment:
This will all disappear into Sock in short order, though for now this keeps Phoenix as Cowboy only